contractsV5/kyberDAO/KyberDAO.sol

```diff
@@ -0,0 +1,649 @@
+        if (votes.length == 0) {
+            return (voteCounts, totalVoteCount);
+        }
+        totalVoteCount = votes[0];
+        voteCounts = new uint[](votes.length - 1);
+        for (uint i = 0; i < voteCounts.length; i++) {
+            voteCounts[i] = votes[i + 1];
+        }
+    }
+
+    function getCampaignOptionVoteCount(uint campID, uint optionID)
+        public view
+        returns(uint voteCount, uint totalVoteCount)
+    {
+        uint[] memory voteCounts = campOptionPoints[campID];
+        if (voteCounts.length == 0 || optionID == 0 || optionID >= voteCounts.length) {
+            return (voteCount, totalVoteCount);
+        }
+        voteCount = voteCounts[optionID];
+        totalVoteCount = voteCounts[0];
+    }
+
+    function getCampaignWinningOptionAndValue(uint campID)
+        public view
+        returns(uint optionID, uint value)
+    {
+        if (!isCampExisted[campID]) { return (0, 0); } // not existed
+
+        Campaign storage camp = campaignData[campID];
+
+        // not found or not ended yet, return 0 as winning option
+        if (camp.endBlock == 0 || camp.endBlock > block.number) { return (0, 0); }
+
+        bool hasConcluded;
+        (hasConcluded, optionID) = decodeWinningOptionData(winningOptionData[campID]);
+        if (hasConcluded) {
+            if (optionID == 0 || optionID >= camp.options.length) {
+                // no winning option or invalid winning option
+                return (0, 0);
+            }
+            return (optionID, camp.options[optionID]);
+        }
+
+        uint totalSupply = camp.totalKNCSupply;
+        // no one has voted in this epoch
+        if (totalSupply == 0) { return (0, 0); }
+
+        uint[] memory voteCounts = campOptionPoints[campID];
+
+        // Finding option with most votes
+        uint winningOption = 0;
+        uint maxVotedCount = 0;
+        for (uint i = 1; i < voteCounts.length; i++) {
+            if (voteCounts[i] > maxVotedCount) {
+                winningOption = i;
+                maxVotedCount = voteCounts[i];
+            } else if (voteCounts[i] == maxVotedCount) {
+                winningOption = 0;
+            }
+        }
+        // more than 1 options have same vote count
+        if (winningOption == 0) { return (0, 0); }
+
+        FormulaData memory formulaData = decodeFormulaParams(camp.formulaParams);
+
+        uint totalVotes = voteCounts[0];
+        // compute voted percentage (in precision)
+        uint votedPercentage = totalVotes.mul(PRECISION).div(camp.totalKNCSupply);
+
+        if (formulaData.minPercentageInPrecision > votedPercentage) { return (0, 0); }
+
+        uint x = formulaData.tInPrecision.mul(votedPercentage).div(PRECISION);
+        if (x <= formulaData.cInPrecision) {
+            // threshold is not negative, need to compare with voted count
+            uint y = formulaData.cInPrecision.sub(x);
+            if (maxVotedCount.mul(PRECISION) < y.mul(totalVotes)) { return (0, 0); }
```

Comment:
In the Katalyst & DAO meeting notes from December
you can find it under "5. Mechanisms to conclude a voting campaign"

Reply:
@Anyhowclick 
please review this code

Reply:
## Quick Primer:
A line can be expressed as `Y = mX + C`. `m` is the gradient, `C` shifts the line vertically up/down.

## Variables
`votedPercentage` = % of votes out of total KNC supply.
`minThreshold` = percentage (of all votes for the campaign) needed for winning option to pass.

## Explanation
We use `minThreshold = - t * votedPercentage + C`, where `t and C >= 0`. So votedPercentage is linearly inversely proportional to minThreshold if votedPercentage exists in [minPercentage, C]

Rearranging, `minThreshold = C - t * votedPercentage`.
First `x = t*votedPercentage`.
If C >= x, `y = C - x = minThreshold >= 0`, so we need to check further (explained below). Otherwise, enough people have voted and we can accept the winning option.

Because not enough people have voted, we need to check if the winning option has crossed the minThreshold. `y` (ie. threshold) is computed in L454, and the check is done in L455, where `maxVotedCount` is the count of winning option votes, and `totalVotes` is votes received for that campaign.

## Conclusion
Based on what I understand, the code seems correct hahaha.

Reply:
haha thanks :))

Reply:
Saving my comment for documentation hahaha